src/library/scala/concurrent/impl/Promise.scala

```diff
@@ -332,77 +387,168 @@ private[concurrent] object Promise {
+    /* Submits the callback function(s) represented by this Callback to be executed with the given value `v` */
+    def submitWithValue(v: Try[T @uncheckedVariance]): Unit
+  }
+
+  /* Represents 0 Callbacks, is used as an initial, sentinel, value for DefaultPromise
+   * This used to be a `case object` but in order to keep `Callbacks`'s methods bimorphic it was reencoded as `val`
+   */
+  final val NoopCallback: Callbacks[Nothing] =
+    new Callback[Nothing](new ExecutionContext {
+      override def execute(r: Runnable): Unit = throw new IllegalStateException("Noop ExecutionContext.execute!")
+      override def reportFailure(t: Throwable): Unit = t.printStackTrace(System.err)
+      override def toString: String = "<Noop>"
+    }, _ => ())
+
+  /* Represents a single Callback function.
+     Precondition: `executor` is prepar()-ed */
+  final class Callback[T](
+    final val executor: ExecutionContext,
+    final val onComplete: Try[T] => Any) extends Callbacks[T] with Runnable with OnCompleteRunnable {
+
+    private[this] final var value: AnyRef = executor // value is initially the EC
+
+    override final def run(): Unit = {
+      val v = value
+      if (v.isInstanceOf[Try[T]]) {
+        value = null
+        try onComplete(v.asInstanceOf[Try[T]]) catch { case NonFatal(e) => executor reportFailure e }
+      } else if (v eq executor) {
+        throw new IllegalStateException("Callback value must be set when running")
+      } else /*if (v eq null)*/ ()
+    }
 
-    private[this] sealed trait Kept[T] extends Promise[T] {
-      def result: Try[T]
+    override final def submitWithValue(v: Try[T]): Unit = 
+      if (this ne NoopCallback) {
+        val e = value
+        if (e eq executor) {
+          value = v // Safe publication of `value`, to run(), is achieved via `executor.execute(this)`
+          // Note that we cannot prepare the ExecutionContext at this point, since we might already be running on a different thread!            
+          try executor.execute(this) catch { case NonFatal(t) => executor reportFailure t }
+        } // else … already submitted or already executed
+      }
+
+    override final def prepend[U >: T](c: Callbacks[U]): Callbacks[U] = {
+      if (c.isInstanceOf[Callback[T]]) {
+        if (c eq NoopCallback) this
+        else if (this eq NoopCallback) c
+        else if (this eq c) this
+        else ManyCallbacks.two(this, c)
+      } else /*if (c.isInstanceOf[ManyCallbacks[T]])*/ {
+        c.asInstanceOf[ManyCallbacks[T]] append this // m append this == this prepend m
+      }
+    }
 
-      override def value: Option[Try[T]] = Some(result)
+    override def toString: String = s"Callback($executor, $onComplete)"
+  }
 
-      override def isCompleted: Boolean = true
+  final object ManyCallbacks {
+    final def two[U](second: Callbacks[U], first: Callbacks[U]): ManyCallbacks[U] =
+      new ManyCallbacks[U](NoopCallback, NoopCallback, second, first, 2)
```

Comment:
Is there a particular reason why you fill the slots back-to-front rather than the naive front-to-back approach that most people would probably initially expect?

Reply:
@sirthias Yes, I believe so: I am replacing the use of List, which is prepend, so I didn't want to break compat by inadvertently changing the order.

Reply:
Ah. Didn't see that it was a list before.
Makes sense.